lib/dataset/sort.cpp

```diff
@@ -3,45 +3,83 @@
 /// @file
 /// @author Simon Heybrock
 #include "scipp/dataset/sort.h"
-#include "scipp/dataset/groupby.h"
+#include "scipp/core/tag_util.h"
+#include "scipp/dataset/extract.h"
 
 namespace scipp::dataset {
 
+namespace {
+
+constexpr auto nan_sensitive_less = [](const auto &a, const auto &b) {
+  if constexpr (std::is_floating_point_v<std::decay_t<decltype(a)>>)
+    if (std::isnan(b))
+      return !std::isnan(a);
+  return a < b;
+};
+
+template <class T> struct IndicesForSorting {
+  static Variable apply(const Variable &key, const SortOrder order) {
+    const auto size = key.dims()[key.dim()];
+    const auto values = key.values<T>();
+    std::vector<std::pair<T, scipp::index>> pairs;
+    pairs.reserve(size);
+    scipp::index i = 0;
+    for (const auto &value : key.values<T>())
+      pairs.emplace_back(value, i++);
+
+    if (order == SortOrder::Ascending)
+      std::sort(pairs.begin(), pairs.end(), [](const auto &a, const auto &b) {
+        return nan_sensitive_less(a.first, b.first);
+      });
+    else
+      std::sort(pairs.begin(), pairs.end(), [](const auto &a, const auto &b) {
+        return nan_sensitive_less(b.first, a.first);
+      });
+
+    auto indices =
+        makeVariable<scipp::index_pair>(Dims{key.dim()}, Shape{size});
+    scipp::index j = 0;
+    for (auto &index : indices.values<scipp::index_pair>().as_span()) {
+      const auto start = pairs[j++].second;
+      index = {start, start + 1};
+    }
```

Comment:
Can this be done with `std::transform`? Took me a moment to understand the loop.